Join-String cmdlet for creating text from pipeline input

powercode opened the pull request:
## PR Summary

A join-object cmdlet that joins pipeline input to text.

#6697 

As requested by @BrucePay.

Final syntax of the cmdlet:
```
Join-String [[-Property] <pspropertyexpression>] [[-Separator] <string>] [-OutputPrefix <string>] [-OutputSuffix <string>] [-UseCulture] [-InputObject <psobject>] [<CommonParameters>]

Join-String [[-Property] <pspropertyexpression>] [[-Separator] <string>] [-OutputPrefix <string>] [-OutputSuffix <string>] [-SingleQuote] [-UseCulture] [-InputObject <psobject>] [<CommonParameters>]

Join-String [[-Property] <pspropertyexpression>] [[-Separator] <string>] [-OutputPrefix <string>] [-OutputSuffix <string>] [-DoubleQuote] [-UseCulture] [-InputObject <psobject>] [<CommonParameters>]

Join-String [[-Property] <pspropertyexpression>] [[-Separator] <string>] [-OutputPrefix <string>] [-OutputSuffix <string>] [-FormatString <string>] [-UseCulture] [-InputObject <psobject>] [<CommonParameters>]
```
```PowerShell
1..3 | Join-String -OutputPrefix "A " -OutputSuffix " B" -Separator "," -SingleQuote | Should -BeExactly "A '1','2','3' B" 
gci | Join-String Name -Separator "," -DoubleQuote
gci | Join-String {$_.Name * 2} "; "
```

## PR Checklist

- [x] [PR has a meaningful title](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
    - Use the present tense and imperative mood when describing your changes
- [x] [Summarized changes](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [ ] [Change is not breaking](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#making-breaking-changes)
- [x] [Make sure all `.h`, `.cpp`, `.cs`, `.ps1` and `.psm1` files have the correct copyright header](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [x] This PR is ready to merge and is not [Work in Progress](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---work-in-progress).
    - If the PR is work in progress, please add the prefix `WIP:` to the beginning of the title and remove the prefix when the PR is ready.
- **User-facing changes**
    - [ ] Not Applicable
    - **OR**
    - [x] User-facing [Documentation needed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
        - [ ] Issue filed - Issue link:
- **Testing - New and feature**
    - [ ] Not Applicable or can only be tested interactively
    - **OR**
    - [x] [Make sure you've added a new test if existing tests do not effectively test the code changed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#before-submitting)
        - [ ] [Add `[feature]` if the change is significant or affects feature tests](https://github.com/PowerShell/PowerShell/blob/master/docs/testing-guidelines/testing-guidelines.md#requesting-additional-tests-for-a-pr)

powercode committed:
Fix disposing of powershell in TypeInferenceContext if it was created internally.

powercode committed:
Adding a PropertyNameCompleter custom completer.

powercode committed:
Adding Join-Object command

powercode committed:
Adding tests

iSazonov commented:
'Object' in the cmdlet name implies that we can join objects of any type - array + array, array + hash, hash + hash, strings and so on.

powercode commented:
It is intended to be the pipeline equivalent of the `-join` operator. @BrucePay  suggested the name `Join-Object` in the issue. `Convert` would imply for me that it converted each item to a string. This cmdlet join them together with a delimiter.

RichardSiddaway commented:
If its going to be the pipeline equivalent of -join wouldn't Join-String be a more descriptive name.  Shouldn't there also be a Split-? cmdlet to match the -split operator

powercode commented:
It depends. It can join objects, or the properties of objects. 

The output is a string, but it works with object input. Just like Group-Object and Sort-Object.

PaulHigin commented on src/System.Management.Automation/engine/CommandCompletion/CompletionCompleters.cs:
```diff
@@ -5126,7 +5126,7 @@ private static bool IsInDscContext(ExpressionAst expression)
         }
 
-        private static void CompleteMemberByInferredType(CompletionContext context, IEnumerable<PSTypeName> inferredTypes, List<CompletionResult> results, string memberName, Func<object, bool> filter, bool isStatic)
+        internal static void CompleteMemberByInferredType(TypeInferenceContext context, IEnumerable<PSTypeName> inferredTypes, List<CompletionResult> results, string memberName, Func<object, bool> filter, bool isStatic)
```
Do we know that CompletionContext.Helper is the same thing has TypeInferenceContext.Helper?

powercode replied:
Only by reading the source. But it was me who refactored the helper into a shared class when I wrote the TypeInferenceContext class.

PaulHigin commented on src/System.Management.Automation/engine/CommandCompletion/CompletionCompleters.cs:
```diff
@@ -6905,4 +6905,83 @@ public object VisitParenExpression(ParenExpressionAst parenExpressionAst)
+    /// <summary>
+    /// Completes with the property names of the InputObject.
+    /// </summary>
+    public class PropertyNameCompleter : IArgumentCompleter
```
I think this should be internal access.

powercode replied:
I understand from the perspective of this CL, but in general, isn't it cool if we could get completion for PropertyNames for third party cmdlets? 
Others (like me :) ) can make use of that too.

iSazonov replied:
I suggest to convert to internal for fast merging the PR and open a tracking issue to wait PowerShell Committee approval.

powercode replied:
It is now internal.

PaulHigin commented on src/System.Management.Automation/engine/CommandCompletion/CompletionCompleters.cs:
```diff
@@ -6905,4 +6905,83 @@ public object VisitParenExpression(ParenExpressionAst parenExpressionAst)
+            }
+
+            int i;
+            for (i = 0; i < pipelineAst.PipelineElements.Count; i++)
```
Have you considered using `Ast.FindAll(...).FirstOrDefault()`?

powercode replied:
Yes. It's a balance. I try hard not to create extra objects, which Linq tends to do. But it sure make the code uglier and harder to read.

And to be honest, I haven't measured the allocations here. 
But it would mean the allocation of a Func delegate, an AstSearcher, a List for the results, and the enumerator. Granted, the last one is an List.Enumerator which is only on the stack.

I fairly often use Linq in higher level code, but when writing infrastructure like this, I try to go with the boring, ugly, but faster option.

PaulHigin commented on src/System.Management.Automation/engine/parser/TypeInferenceVisitor.cs:
```diff
@@ -121,14 +122,16 @@ public int GetHashCode(PSTypeName obj)
         public TypeInferenceContext()
         : this(PowerShell.Create(RunspaceMode.CurrentRunspace))
         {
+            _ownsPowerShell = true;
```
This doesn't look right to me.  This class can own PowerShell/Runspace only if PowerShell is created with `Runspace.NewRunspace` mode.  Otherwise you are disposing a runspace created by someone else (and possibly the thread default runspace).

powercode replied:
Reverting the whole file.

powercode commented on src/Microsoft.PowerShell.Commands.Utility/commands/utility/join-object.cs:
```diff
@@ -0,0 +1,211 @@
+    public class JoinObjectCommand : PSCmdlet
+    {
+        // ReSharper disable once CollectionNeverQueried.Local
+        private readonly List<PSObject> _inputObjects = new List<PSObject>(50);
```
It is cheaper to allocate a list that is a little larger than to reallocate and copy when out-grown.
But the size can always be argued. 

When I profile memory allocations, Array allocations on List.Resize is not uncommon, and when I see them, I try to find a sensible default size do get rid of the most common once.

powercode commented on src/Microsoft.PowerShell.Commands.Utility/commands/utility/join-object.cs:
```diff
@@ -0,0 +1,211 @@
+namespace Microsoft.PowerShell.Commands.Utility
+{
+    /// <summary>
+    /// Group-Object implementation.
```
oops :)

powercode committed:
Reverting incorrect dispose in TypeInferenceVisitor

powercode commented on src/Microsoft.PowerShell.Commands.Utility/commands/utility/join-object.cs:
```diff
@@ -0,0 +1,211 @@
+    /// </summary>
+    [Cmdlet(VerbsCommon.Join, "Object", RemotingCapability = RemotingCapability.None, DefaultParameterSetName = "default")]
+    [OutputType(typeof(string))]
+    public class JoinObjectCommand : PSCmdlet
```
yes

powercode committed:
Addressing Pauls review comments.

powercode commented:
@PaulHigin Thx for the review.

powercode commented on src/Microsoft.PowerShell.Commands.Utility/commands/utility/join-object.cs:
```diff
@@ -0,0 +1,211 @@
+
+            if (PropertyName == null)
+            {
+                if (_inputObjects.Count > 0)
```
Yes, much better!

powercode committed:
Early out when inputobject count == 0.

iSazonov commented on test/powershell/Modules/Microsoft.PowerShell.Utility/Join-Object.Tests.ps1:
```diff
@@ -0,0 +1,87 @@
+# Copyright (c) Microsoft Corporation. All rights reserved.
+# Licensed under the MIT License.
+Describe "Join-Object" -Tags "CI" {
```
Please add new line.

iSazonov commented on src/Microsoft.PowerShell.Commands.Utility/commands/utility/join-object.cs:
```diff
@@ -0,0 +1,211 @@
+        /// </summary>
+        [Parameter(Position = 0)]
+        [ArgumentCompleter(typeof(PropertyNameCompleter))]
+        public object PropertyName { get; set; }
```
Can we use PSPropertyExpression type like in #6934?

powercode replied:
PropertyExpression has some extra semantics, handling hashtables like `@{n="x";ex={2,3,4}}`.

I looked at it, but was not convinced it was a fit. But there's definitely an overlap. Input here is most welcome.

powercode commented on src/Microsoft.PowerShell.Commands.Utility/commands/utility/join-object.cs:
```diff
@@ -0,0 +1,211 @@
+            CommandAst commandAst,
+            IDictionary fakeBoundParameters)
+        {
+            var res = new List<CompletionResult>(10);
```
Then you end up with races if used from multiple runspaces.

iSazonov replied:
The list looks as a const for current session. Do I skip something?

powercode replied:
Yes, it is captured by the local function `AddMatching`.
So the result is depending on `wordToComplete`.

iSazonov replied:
I see now. :-) It is Friday. I can not catch up with my idea :-)
I am looking at 'new CompletionResult()'. If they isn't modifed later we could make it static and only reference in target List?

iSazonov commented on src/Microsoft.PowerShell.Commands.Utility/commands/utility/join-object.cs:
```diff
@@ -0,0 +1,211 @@
+                    case "\r": return "`r";
+                    case "\n": return "`n";
+                    case "\r\n": return "`r`n";
+                    default: return Environment.NewLine.Replace("\r", "`r").Replace("\n", "`n");
```
Seems the switch is superfluous. We could leave only
``'c#
return Environment.NewLine.Replace("\r", "`r").Replace("\n", "`n"); 
```

powercode replied:
Yes it could. I wrote it the way I did to reduce allocated objects. May not be worth it. 
The string literals will be loaded from metadata and interned. There will only exist one instance of them. The default line will create two new strings each time.

iSazonov replied:
Environment.NewLine is static. So I think we should use the usual method
```c#
public static string NewLineText
{
    get
    {
#if UNIX
        return "`n"
#else
        return "`r`n"
#endif
    }
}
```

powercode replied:
Good point!

powercode replied:
But I think MacOS uses ``` `r```

iSazonov replied:
No, only very old MacOs versions. See my comment below.

powercode reviewed (commented):
A general question is how we should handle `null` and `AutomationNull.Value` in the input?

iSazonov commented:
`AutomationNull.Value` is ignored by default in pipes 
https://github.com/PowerShell/PowerShell/blob/4fc784e840b8df00d2552045c6ed52d027659d2b/src/System.Management.Automation/engine/Pipe.cs#L380

Ex.:
```powershell
$null,$null | % { 1 }
1
1

[System.Management.Automation.Internal.AutomationNull]::Value,[System.Management.Automation.Internal.AutomationNull]::Value | % { 1 }
```

PaulHigin reviewed (approved):
LGTM

vexx32 commented:
I'd be inclined to call it Join-String instead, since nothing of the original objects are preserved (counter to things like Group-Object or Select-Object)

powercode committed:
Renaming Join-Object => Join-String.

PreScript to Prefix
PostScript to Suffix.
PropertyName to Property.

powercode committed:
Fixing missing TypeInferenceVisitor Dispose change.

powercode committed:
newline at eof.

powercode committed:
Renaming files *-object -> *-string

powercode committed:
platform specific newline completion text.

powercode committed:
Ignore null and automationnull input.

powercode committed:
Moving the early exit to the top of the function.

powercode committed:
Codefactor fixes.

powercode committed:
Adding Join-String to defaultcommands test.

iSazonov commented on src/Microsoft.PowerShell.Commands.Utility/commands/utility/join-string.cs:
```diff
@@ -0,0 +1,210 @@
+        /// Gets or sets if the output items should we wrapped in single quotes.
+        /// </summary>
+        [Parameter(ParameterSetName = "Quote")]
+        public SwitchParameter Quote { get; set; }
```
SingleQuote supplements DoubleQuote.

Another thought is to make one parameter (which replace both Quote and DoubleQuote parameters) with allowed values SingleQuote and DoubleQuote. Seems better UX. This also reduce account of parameter sets.

powercode replied:
I don't. I have used this cmdlet in a slightly different but very similar implementation, and the most common use case, by far, is to create strings for use in code.
Having `DoubleQuote` and `SingleQuote` is so much nicer than to have to specify a `Format`. 

I can for sure see the value of Format, but it is not as nice from an interactive perspective as the switches.

iSazonov commented on src/Microsoft.PowerShell.Commands.Utility/commands/utility/join-string.cs:
```diff
@@ -0,0 +1,210 @@
+        /// Gets or sets the input object to join into text.
+        /// </summary>
+        [Parameter(ValueFromPipeline = true, Mandatory = true)]
+        public PSObject InputObject { get; set; }
```
Why only PSObject? Maybe object?

iSazonov commented on src/Microsoft.PowerShell.Commands.Utility/commands/utility/join-string.cs:
```diff
@@ -0,0 +1,210 @@
+        /// <inheritdoc />
+        protected override void ProcessRecord()
+        {
+            if (InputObject != null && InputObject != AutomationNull.Value)
```
Make sense remove `InputObject != AutomationNull.Value`?

iSazonov commented on src/Microsoft.PowerShell.Commands.Utility/commands/utility/join-string.cs:
```diff
@@ -0,0 +1,210 @@
+        {
+            if (_inputObjects.Count == 0)
+            {
+                return;
```
Should we return null or empty string?

iSazonov commented on src/Microsoft.PowerShell.Commands.Utility/commands/utility/join-string.cs:
```diff
@@ -0,0 +1,210 @@
+                for (var index = 1; index < _inputObjects.Count; index++)
+                {
+                    builder.Append(Delimiter);
+                    AppendValue(LanguagePrimitives.ConvertTo<string>(_inputObjects[index]));
```
we do not need to collect input objects - we need only result strings. So we should move the ConvertTo to the ProcessRecord() method.

iSazonov commented on src/Microsoft.PowerShell.Commands.Utility/commands/utility/join-string.cs:
```diff
@@ -0,0 +1,210 @@
+            }
+            else
+            {
+                throw new ArgumentException();
```
We should return useful error message. Or yet use PSPropertyExpression.

iSazonov commented on src/Microsoft.PowerShell.Commands.Utility/commands/utility/join-string.cs:
```diff
@@ -0,0 +1,210 @@
+            get
+            {
+#if UNIX
+                return Platform.IsMacOS ? "`r" : "`n";
```
It is only for very old MacOs versions. Please remove.
CoreFX doesn't use this https://source.dot.net/#System.Private.CoreLib/src/System/Environment.cs,261
For reference https://en.wikipedia.org/wiki/Newline

iSazonov replied:
Please address the comment.

iSazonov commented on src/System.Management.Automation/engine/CommandCompletion/CompletionCompleters.cs:
```diff
@@ -6905,4 +6906,82 @@ public object VisitParenExpression(ParenExpressionAst parenExpressionAst)
+            IDictionary fakeBoundParameters)
+        {
+            var pipelineAst = commandAst.Parent as PipelineAst;
+            if (pipelineAst == null)
```
Why not C# 7.0 syntax?

iSazonov commented on src/System.Management.Automation/engine/CommandCompletion/CompletionCompleters.cs:
```diff
@@ -6905,4 +6906,82 @@ public object VisitParenExpression(ParenExpressionAst parenExpressionAst)
+                }
+
+                var astPair = pair as AstPair;
+                if (astPair?.Argument == null)
```
Why not C# 7.0 syntax? Looks more readable.

iSazonov commented on test/powershell/Modules/Microsoft.PowerShell.Utility/Join-String.Tests.ps1:
```diff
@@ -0,0 +1,99 @@
+    It "Should return a single string" {
+        $actual = $testObject | Join-String
+
+        $actual.Count       | Should -Be 1
```
Extra spaces.

iSazonov commented on test/powershell/Modules/Microsoft.PowerShell.Utility/Join-String.Tests.ps1:
```diff
@@ -0,0 +1,99 @@
+			$n -in $completionTexts | Should -BeTrue
+		}
+	}
+
```
Please remove the extra line.

iSazonov commented on test/powershell/engine/Basic/DefaultCommands.Tests.ps1:
```diff
@@ -319,6 +319,7 @@ Describe "Verify approved aliases list" -Tags "CI" {
 "Cmdlet", "Invoke-WmiMethod",                                   ,                     $($FullCLR                               )
 "Cmdlet", "Invoke-WSManAction",                                 ,                     $($FullCLR -or $CoreWindows              )
+"Cmdlet", "Join-String",                                        ,                     $(             $CoreWindows -or $CoreUnix)
 "Cmdlet", "Join-Path",                                          ,                     $($FullCLR -or $CoreWindows -or $CoreUnix)
```
Please swap the two lines. Alphabetical order.

powercode committed:
sorting cmdlet names.

powercode commented on src/Microsoft.PowerShell.Commands.Utility/commands/utility/join-string.cs:
```diff
@@ -0,0 +1,210 @@
+        [Parameter(Position = 1)]
+        [ArgumentCompleter(typeof(JoinItemCompleter))]
+        [AllowEmptyString]
+        public string Delimiter { get; set; }
```
GetVariableValue is not available in a static context.

powercode commented on src/Microsoft.PowerShell.Commands.Utility/commands/utility/join-string.cs:
```diff
@@ -0,0 +1,210 @@
+
+            if (Delimiter == null)
+            {
+                Delimiter = LanguagePrimitives.ConvertTo<string>(GetVariableValue("OFS"));
```
No, that code makes use of the initialization that the engine sets up after the cmdlet objects constructor has been run.

powercode committed:
Addressing Ilya's feedback

powercode committed:
Making PropertyNameCompleter internal.

powercode commented:
What about aliases?

Join has the alias 'j' and String seems to be 's'

Should there be a 'js' alias?

powercode commented on src/Microsoft.PowerShell.Commands.Utility/commands/utility/join-string.cs:
```diff
@@ -0,0 +1,210 @@
+    /// <summary>
+    /// Join-Object implementation.
+    /// </summary>
+    [Cmdlet(VerbsCommon.Join, "String", RemotingCapability = RemotingCapability.None, DefaultParameterSetName = "default")]
```
I don't know much about this. Who does documentation?

powercode replied:
I don't think the developer who wrote it excels at documentation in English. Rumor has it that he was willing to write it in the language of the Gods and the Heroes, though.

BrucePay reviewed (changes requested):
This code really should be using `PSPropertyExpression`.

iSazonov commented on src/System.Management.Automation/engine/CommandCompletion/CompletionCompleters.cs:
```diff
@@ -6905,4 +6906,80 @@ public object VisitParenExpression(ParenExpressionAst parenExpressionAst)
+    /// <summary>
+    /// Completes with the property names of the InputObject.
+    /// </summary>
+    internal class PropertyNameCompleter : IArgumentCompleter
```
Is the cmdlet already approved by PowerShell Committee? If so we can merge "as is" and consider this api later.

powercode replied:
If you are gonna discuss that, please consider making AstTypeInference public too. I need it from EditorServices.

powercode replied:
I had it public in my initial commit, as I believed it to be generally useful. Changed it to internal after @PaulHigin feedback.

iSazonov commented on src/Microsoft.PowerShell.Commands.Utility/commands/utility/join-string.cs:
```diff
@@ -0,0 +1,211 @@
+            var quoteChar = Quote ? '\'' : DoubleQuote ? '"' : char.MinValue;
+
+            const int DefaultOutputStringCapacity = 256;
+            var builder = new StringBuilder(DefaultOutputStringCapacity);
```
We should do the join in ProcessRecord() too and here only add suffix (prefix in BeginProcessing()).

powercode commented on src/Modules/Unix/Microsoft.PowerShell.Utility/Microsoft.PowerShell.Utility.psd1:
```diff
@@ -16,7 +16,7 @@ CmdletsToExport= "Format-List", "Format-Custom", "Format-Table", "Format-Wide",
     "Select-String", "Measure-Object", "New-Alias", "New-TimeSpan", "Read-Host", "Set-Alias", "Set-Date",
     "Start-Sleep", "Tee-Object", "Measure-Command", "Update-TypeData", "Update-FormatData",
-    "Remove-TypeData", "Get-TypeData", "Write-Host", "Write-Progress", "New-Object", "Select-Object",
+    "Remove-TypeData", "Get-TypeData", "Write-Host", "Write-Progress", "New-Object", "Select-Object", "Join-String",
```
I got lazy when I renamed it from Join-Object. Well, laz*ier* :)

powercode commented on test/powershell/Modules/Microsoft.PowerShell.Utility/Join-String.Tests.ps1:
```diff
@@ -0,0 +1,98 @@
+    }
+
+    It "Should join property values with default delimiter" {
+        $expected = $testObject.Name -join $ofs
```
Does it make sense to allow wildcards, but not have an array?

powercode replied:
I don't get the usecase for the wildcard here. Can you give an example when you think it would be valuable?

powercode commented on test/powershell/Modules/Microsoft.PowerShell.Utility/Join-String.Tests.ps1:
```diff
@@ -0,0 +1,98 @@
+        $actual | Should -BeExactly $expected
+    }
+
+    It "Should join property values DoubleQuoted" {
```
I like that idea!

powercode replied:
Do we need an IFormatProvider/Culture too?

powercode committed:
Addressing most of Bruce's feedback.

powercode commented:
I have addressed most of the feedback.

What I know is remaining is the following:

- Should Property be an Array? If so, why?
- Should Wildcards be allowed? When is this useful?
- Do we need a `-Culture` property for use with format, or should we just use the current culture?
- Should we have an alias for the cmdlet?
- Docs in other languages than Swedish.

iSazonov reviewed (commented):
> Should Property be an Array? If so, why?

One scenario I see instantly - very flexible convert to csv.

> Do we need a -Culture property for use with format, or should we just use the current culture?

In Get-Date we have Format parameter without `Cuture`.

> Should we have an alias for the cmdlet?

I believe no - we should delegate this users.

powercode committed:
Moving OutputPrefix to BeginProcessing

iSazonov commented on src/Microsoft.PowerShell.Commands.Utility/commands/utility/join-string.cs:
```diff
@@ -0,0 +1,210 @@
+
+                if (_firstInputObject)
+                {
+                    _firstInputObject = false;
```
Now we do not need this

powercode replied:
Well, we still need a way to determine if we should skip the first delimiter, don't we?

iSazonov replied:
Yes, I forget about the separator.

powercode commented:
None of the test failures are related to my commits.

anmenaga commented:
I restarted failed test runs; we'll see how they go...

anmenaga commented:
@BrucePay can you please take a look again at this PR? Thanks.

iSazonov commented on src/Microsoft.PowerShell.Commands.Utility/commands/utility/join-string.cs:
```diff
@@ -0,0 +1,210 @@
+
+        /// <summary>
+        /// Gets or sets a format string that is applied to each input object.
+        /// </summary>
```
I'd expand the comment with what is the format and/or add a link on docs.

iSazonov commented on src/Microsoft.PowerShell.Commands.Utility/commands/utility/join-string.cs:
```diff
@@ -0,0 +1,210 @@
+    [OutputType(typeof(string))]
+    public sealed class JoinStringCommand : PSCmdlet
+    {
+        /// <summary>A bigger default to not get re-allocations in common use cases.</summary>
```
Please use file pattern:
```c#
/// <Summary>
/// A bigger default to not get re-allocations in common use cases.
/// </Summary> 
```

iSazonov commented on src/Microsoft.PowerShell.Commands.Utility/commands/utility/join-string.cs:
```diff
@@ -0,0 +1,210 @@
+                {
+                    _outputBuilder.Append(_quoteChar);
+                    _outputBuilder.Append(stringValue);
+                    _outputBuilder.Append(_quoteChar);
```
We could use
```c#
_outputBuilder.Append(_quoteChar).Append(stringValue).(_quoteChar); 
```

iSazonov commented on src/Microsoft.PowerShell.Commands.Utility/commands/utility/join-string.cs:
```diff
@@ -0,0 +1,210 @@
+                }
+                else
+                {
+                    _outputBuilder.AppendFormat(CultureInfo.CurrentCulture, FormatString, stringValue);
```
I'd want to get confirmation that this should be `CurrentCulture` not `InvariantCulture`.
And maybe discuss `-FormatCulture` parameter.

powercode replied:
Same here. Now sure what is right here.

mklement0 replied:
PowerShell consistently uses the invariant culture in to/from-string conversions, so it should probably be `InvariantCulture`, at least by default, perhaps with an optional  `-UseCulture` switch (analogous to the `*-Csv` cmdlets).

iSazonov replied:
```powershell
 gcm -ParameterName Culture

CommandType     Name                                               Version    Source
-----------     ----                                               -------    ------
Cmdlet          Compare-Object                                     3.1.0.0    Microsoft.PowerShell.Utility
Cmdlet          Group-Object                                       3.1.0.0    Microsoft.PowerShell.Utility
Cmdlet          New-PSSessionOption                                6.0.0.0    Microsoft.PowerShell.Core
Cmdlet          Sort-Object                                        3.1.0.0    Microsoft.PowerShell.Utility
```

We already have `-Culture` parameter. Make sense re-use the name?

mklement0 replied:
@iSazonov: 

I think `-UseCulture` is the correct name, after all:

* The `-Culture <string>` parameters you reference:
    * require an _argument_, namely a _given_ culture
    * are used with cmdlets that are current-culture-_sensitive_ by default, to allow opt-in to a _different_ culture.

* The `-UseCulture` switch, by contrast:
    * is used with cmdlets that are culture-_invariant_ by default, to allow opt-in to the _current_ culture (currently applies only to the `*-Csv` cmdlets).

The latter is what is being implemented here.

iSazonov replied:
@mklement0 Thanks! We need discuss Select-String  too.

iSazonov commented on test/powershell/Modules/Microsoft.PowerShell.Utility/Join-String.Tests.ps1:
```diff
@@ -0,0 +1,118 @@
+
+    It "Should be called using an object as piped without error with no switches" {
+        {$testObject | Join-String } | Should -Not -Throw
+    }
```
We could remove the test - any test below do the test.

BrucePay reviewed (approved):
LGTM

powercode committed:
Addressing Ilya's comments.

anmenaga commented:
@powercode Looks like this PR still needs a couple of minor updates that are mentioned in @iSazonov's review; especially InvariantCulture issue.  Otherwise it seems close to being done.

powercode committed:
Adding UseCulture switch to the Format parameter set.

powercode commented:
I just realized that when joining dates, I always get the "US", Invariant formatting, which is literally never the output a Swede would want. 

To get the `UseCulture` switch to have effect, I need to do some minor changes in how LanguagePrimitives makes the conversion to strings, and provide an overload that accepts an IFormatProvider.

powercode committed:
Make TryConvertTo<string> respect the passed formatting provider.

daxian-dbw commented:
The last commit (`"Make TryConvertTo<string> respect the passed formatting provider."`) looks good to me.

iSazonov commented:
What is current policy: don't merge until doc issue is created?
